eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -554,89 +558,144 @@ instance Semigroup ScriptResult where
-          <! D (decodeCostModel PlutusV2)
-          <! From
-          <! From
-          <! D (decodeList (fromPlainDecoder Cborg.decode))
-          <! From
-      dec n = Invalid n
-
-debugPlutus :: Version -> String -> PlutusDebugInfo
+debugPlutus ::
+  Version ->
+  String ->
+  PlutusDebugInfo
 debugPlutus version db =
   case B64.decode (BSU.fromString db) of
     Left e -> DebugBadHex (show e)
     Right bs ->
-      case decodeFull' version bs of
-        Left e -> DebugCannotDecode (show e)
-        Right pdb@(PlutusDebugV1 cm units script ds pv) ->
-          case PV1.evaluateScriptRestricting
-            (transProtocolVersion pv)
-            PV1.Verbose
-            (getEvaluationContext cm)
-            (transExUnits units)
-            script
-            ds of
-            (logs, Left e) -> DebugInfo logs (PlutusErrorV1 e) pdb
-            (_, Right ex) -> DebugSuccess ex
-        Right pdb@(PlutusDebugV2 cm units script ds pv) ->
-          case PV2.evaluateScriptRestricting
-            (transProtocolVersion pv)
-            PV2.Verbose
-            (getEvaluationContext cm)
-            (transExUnits units)
-            script
-            ds of
-            (logs, Left e) -> DebugInfo logs (PlutusErrorV2 e) pdb
-            (_, Right ex) -> DebugSuccess ex
+      let plutusDebugLangDecoder ::
+            (Typeable l, IsLanguage l) =>
+            Either DecoderError (PlutusDebugLang l)
+          plutusDebugLangDecoder = decodeFull' version bs
+          plutusDebugDecoder =
+            case plutusDebugLangDecoder @'PlutusV1 of
+              Right plutusDebug -> pure $ PlutusDebug plutusDebug
+              Left _ ->
+                case plutusDebugLangDecoder @'PlutusV2 of
+                  Right plutusDebug -> pure $ PlutusDebug plutusDebug
+                  Left err -> Left err
+       in case plutusDebugDecoder of
+            Left e -> DebugCannotDecode (show e)
+            Right (PlutusDebug (PlutusDebugLang sl costModel exUnits scripts pData protVer)) ->
+              let evalV1 =
+                    PV1.evaluateScriptRestricting
+                      (transProtocolVersion protVer)
+                      PV1.Verbose
+                      (getEvaluationContext costModel)
+                      (transExUnits exUnits)
+                      scripts
+                  evalV2 =
+                    PV2.evaluateScriptRestricting
+                      (transProtocolVersion protVer)
+                      PV2.Verbose
+                      (getEvaluationContext costModel)
+                      (transExUnits exUnits)
+                      scripts
+               in case sl of
+                    SPlutusV1 ->
+                      case evalV1 $ unPlutusData pData of
+                        (logs, Left e) ->
+                          DebugInfo logs (PlutusErrorV1 e) $
+                            PlutusDebug $
+                              PlutusDebugLang sl costModel exUnits scripts pData protVer
```

Comment:
@lehins here, GHC complained about now figuring out what the `(l :: Language)` was supposed to be, if the `PlutusDebug` construction was factored out within the `let`. So I had to make it this way. Even making a function `plutusDebug :: SLanguage l -> PlutusDebug` in the `let` raises the same error.

Reply:
Yeah, it's fine. I can see how ghc would be confused about the `l` here.